AFQ/api.py

```diff
@@ -71,135 +72,216 @@ def do_preprocessing():
-            elif name == "SLF":
-                for hemi in ['_R', '_L']:
-                    afq_bundles[name + hemi] = {
-                        'ROIs': [templates[name + '_roi1' + hemi],
-                                 templates[name + '_roi2' + hemi],
-                                 templates["SLFt_roi2" + hemi]],
+        Parameters
+        ----------
+        bundle_names : list, optional
+            A list of the bundles to be used in this case. Default: all of them
+
+        seg_algo: One of {"afq", "reco", "reco16", "reco80"}
+            The bundle segmentation algorithm to use.
+                "afq" : Use waypoint ROIs + probability maps, as described
+                in [Yeatman2012]_
+                "reco" / "reco16" : Use Recobundles [Garyfallidis2017]_
+                with a 16-bundle set.
+                "reco80": Use Recobundles with an 80-bundle set.
+
+        resample_to : Nifti1Image or bool, optional
+            If set, templates will be resampled to the affine and shape of this
+            image. If False, no resampling will be done.
+            Default: False
+        """
+        if not (isinstance(bundle_info, dict)
+                or isinstance(bundle_info, list)):
+            raise TypeError((
+                f"bundle_info must be a dict or a list,"
+                f" currently a {type(bundle_info)}"))
+
+        if isinstance(bundle_info, dict):
+            self.bundle_names = list(bundle_info.keys())
+            self._dict = bundle_info.copy()
+        else:
+            self.bundle_names = bundle_info.copy()
+            self._dict = {}
+        self.seg_algo = seg_algo.lower()
+
+        expanded_bundle_names = []
+        for bundle_name in self.bundle_names:
+            if self.seg_algo == "afq":
+                if bundle_name in ["FA", "FP"]\
+                        or bundle_name in CALLOSUM_BUNDLES:
+                    expanded_bundle_names.append(bundle_name)
+                else:
+                    expanded_bundle_names.append(bundle_name + "_R")
+                    expanded_bundle_names.append(bundle_name + "_L")
+            elif self.seg_algo.startswith("reco"):
+                if bundle_name in RECO_UNIQUE\
+                        or bundle_name == "whole_brain":
+                    expanded_bundle_names.append(bundle_name)
+                else:
+                    expanded_bundle_names.append(bundle_name + "_R")
+                    expanded_bundle_names.append(bundle_name + "_L")
+            else:
+                raise ValueError(
+                    "Input: %s is not a valid input`seg_algo`"
+                    % self.seg_algo)
+        self.bundle_names = expanded_bundle_names
+
+        # Each bundles gets a digit identifier
+        # (to be stored in the tractogram)
+        self.uid_dict = {}
+        for ii, b_name in enumerate(self.bundle_names):
+            self.uid_dict[b_name] = ii + 1
+            self.c_uid = ii + 2
```

Comment:
Just making sure that I understand: this is to that when we call `__setitem__` you have an integer ready to give it? Maybe make this a private attribute and add a comment on this line?

Reply:
Yes, I agree